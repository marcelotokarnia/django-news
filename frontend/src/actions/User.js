import {
    REQUEST_USER, RECEIVE_USER,
    RECEIVE_USER_ERROR,
    REQUEST_CHANGE_USER_PREFERENCE,
    RECEIVE_USER_PREFERENCE } from './types'
import { checkStatus, parseJSON } from './utils'
import * as API from './api'

export const requestUser = () => ({
  type: REQUEST_USER,
})

export const receiveUser = profile => ({
  type: RECEIVE_USER,
  profile: profile
})

export const receiveError = error => ({
  type: RECEIVE_USER_ERROR,
  error,
})

export const fetchUser = () => (dispatch) => {
  dispatch(requestUser())

  return window.fetch(`${API.USER}?format=json`, {
    method: 'GET',
    credentials: 'same-origin',
  })
    .then(checkStatus)
    .then(parseJSON)
    .then(profile => dispatch(receiveUser(profile)))
    .catch(error => dispatch(receiveError(error)))
}
