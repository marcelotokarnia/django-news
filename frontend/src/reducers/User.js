import { clone } from 'ramda'
import {
  REQUEST_USER,
  RECEIVE_USER,
  RECEIVE_USER_ERROR,
  REQUEST_CHANGE_USER_PREFERENCE,
  RECEIVE_USER_PREFERENCE
} from '../actions/types'


const initialState = {
  username: '',
  categories: [],
  isFetching: false,
  isMutating: false,
}

const User = (state = initialState, action) => {
  switch (action.type) {
    case REQUEST_USER: {
      return {
        ...state,
        isFetching: true,
      }
    }
    case RECEIVE_USER: {
      return {
        ...state,
        isFetching: false,
        username: action.profile.user.username,
        categories: clone(action.profile.categories)
      }
    }
    case RECEIVE_USER_ERROR: {
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    }
    case REQUEST_CHANGE_USER_PREFERENCE: {
      return {
        ...state,
        isMutating: true,
      }
    }
    case RECEIVE_USER_PREFERENCE: {
      return {
        ...state,
        isMutating: false,
        categories: clone(action.profile.categories),
      }
    }
    default:
      return state
  }
}

export default User
