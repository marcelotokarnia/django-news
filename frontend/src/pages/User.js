import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import { map, assoc, prop, contains, without, append } from 'ramda'
import { mutatePreferences } from '../actions/User'
import Loading from '../components/Loading'

class User extends Component {
  static propTypes = {
    categories: PropTypes.arrayOf(PropTypes.object).isRequired,
    isFetching: PropTypes.bool.isRequired,
    isFetchingUser: PropTypes.bool.isRequired,
    isMutating: PropTypes.bool.isRequired,
    preferred: PropTypes.arrayOf(PropTypes.object).isRequired,
    username: PropTypes.string.isRequired,
    mutatePreferences: PropTypes.func.isRequired,
  }

  constructor({ preferred }) {
    super()
    this.state = {
      preferred,
    }
  }

  componentWillReceiveProps = (nextProps) => {
    if (this.props.isFetchingUser && !nextProps.isFetchingUser) {
      this.setState({ preferred: nextProps.preferred })
    }
  }

  treatCategories = () => {
    const { categories } = this.props
    const { preferred } = this.state
    return map(cat => assoc('isPreferred', contains(prop('name', cat), preferred))(cat), categories)
  }

  handleCategoryClick = (cat) => {
    const { preferred } = this.state
    if (contains(cat, preferred)) {
      this.setState({ preferred: without([cat], preferred) })
    } else {
      this.setState({ preferred: append(cat, preferred) })
    }
  }

  render = () => {
    const { username, isFetching, isMutating } = this.props
    const { preferred } = this.state
    if (isFetching) {
      return <Loading />
    }
    const categories = this.treatCategories()
    return (
      <div className="font-display dark-gray flex flex-wrap justify-center w-100">
        <div className="center w-50-l">
          <h1 className="f2 tc mb5 nt5 dark-gray">WELCOME, <b className="ttu blue">{username}</b></h1>
          <div className="mb4">
            <label htmlFor="id_username" className="w-100 f4">MY INTERESTS</label>
          </div>
          <div className="mb5">
            {map(({
              name, color, isPreferred,
            }) => (
              <span
                onClick={() => this.handleCategoryClick(name)}
                key={name}
                style={{ 'border-color': color, color }}
                className={`${isMutating ? 'o-20' : 'pointer'} f3 mr3 pv2 ph4 ba user-category__radius ${isPreferred && 'bw2 b'}`}
              >
                {name}
              </span>
            ), categories)}
          </div>
          <div className="flex flex-wrap justify-center w-100 mb3">
            <button
              className={`${isMutating && 'o-20'} bg-blue white h3-l h3-m h3-ns h3-xs f3 w-30 bn center`}
              onClick={() => !isMutating && this.props.mutatePreferences(preferred)}
            >
              {isMutating ? 'SAVING' : 'SAVE'}
            </button>
          </div>
          <div className="flex flex-wrap justify-center w-100">
            <Link className="bg-white blue h3-l h3-m h3-ns h3-xs f3 center" to="/news">
              BACK TO HOME
            </Link>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  categories: state.Categories.categories,
  isFetching: state.Categories.isFetching,
  isFetchingUser: state.User.isFetching,
  isMutating: state.User.isMutating,
  preferred: state.User.categories,
  username: state.User.username,
})

const mapDispatchToProps = dispatch => ({
  mutatePreferences: bindActionCreators(mutatePreferences, dispatch),
})

export default connect(mapStateToProps, mapDispatchToProps)(User)
