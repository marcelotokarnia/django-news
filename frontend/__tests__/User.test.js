import React from 'react'
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'
import User from '../src/pages/User'

const mockStore = configureStore()
const dispatch = jest.fn()
const state = {
  Categories: {
    categories: [{ color: '#353535', name: 'POLITICS' }],
    isFetching: false,
  },
  User: {
    isFetching: false,
    isMutating: false,
    categories: [{ name: 'POLITICS' }],
    mutatePreferences: jest.fn(),
    username: 'marcelotokarnia',
  },
}

describe('User component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(<User dispatch={dispatch} store={mockStore({ ...state })} />)
  })

  it('should render correctly', () => {
    expect(wrapper.dive().find('.font-display')).toMatchSnapshot()
  })

  it('should be a div element', () => {
    expect(wrapper.dive().find('.font-display').is('div')).toBe(true)
  })

  it('should not have class o-20 if not mutating', () => {
    expect(wrapper.dive().find('.user-category__radius').hasClass('o-20')).toBe(false)
  })

  it('should have class o-20 if mutating', () => {
    const wrapperMutating = shallow(<User dispatch={dispatch} store={mockStore({ ...state, User: { ...state.User, isMutating: true } })} />)
    expect(wrapperMutating.dive().find('.user-category__radius').hasClass('o-20')).toBe(true)
  })

  it('should render correctly when is mutating', () => {
    const wrapperMutating = shallow(<User dispatch={dispatch} store={mockStore({ ...state, isMutating: true })} />)
    expect(wrapperMutating.dive().find('.font-display')).toMatchSnapshot()
  })
})
