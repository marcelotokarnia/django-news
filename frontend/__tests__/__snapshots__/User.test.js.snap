// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component should render correctly 1`] = `
<div
  className="font-display dark-gray flex flex-wrap justify-center w-100"
>
  <div
    className="center w-50-l"
  >
    <h1
      className="f2 tc mb5 nt5 dark-gray"
    >
      WELCOME, 
      <b
        className="ttu blue"
      >
        marcelotokarnia
      </b>
    </h1>
    <div
      className="mb4"
    >
      <label
        className="w-100 f4"
        htmlFor="id_username"
      >
        MY INTERESTS
      </label>
    </div>
    <div
      className="mb5"
    >
      <span
        className="pointer f3 mr3 pv2 ph4 ba user-category__radius false"
        key="POLITICS"
        onClick={[Function]}
        style={
          Object {
            "border-color": "#353535",
            "color": "#353535",
          }
        }
      >
        POLITICS
      </span>
    </div>
    <div
      className="flex flex-wrap justify-center w-100 mb3"
    >
      <button
        className="false bg-blue white h3-l h3-m h3-ns h3-xs f3 w-30 bn center"
        onClick={[Function]}
      >
        SAVE
      </button>
    </div>
    <div
      className="flex flex-wrap justify-center w-100"
    >
      <Link
        className="bg-white blue h3-l h3-m h3-ns h3-xs f3 center"
        replace={false}
        to="/news"
      >
        BACK TO HOME
      </Link>
    </div>
  </div>
</div>
`;

exports[`User component should render correctly when is mutating 1`] = `
<div
  className="font-display dark-gray flex flex-wrap justify-center w-100"
>
  <div
    className="center w-50-l"
  >
    <h1
      className="f2 tc mb5 nt5 dark-gray"
    >
      WELCOME, 
      <b
        className="ttu blue"
      >
        marcelotokarnia
      </b>
    </h1>
    <div
      className="mb4"
    >
      <label
        className="w-100 f4"
        htmlFor="id_username"
      >
        MY INTERESTS
      </label>
    </div>
    <div
      className="mb5"
    >
      <span
        className="pointer f3 mr3 pv2 ph4 ba user-category__radius false"
        key="POLITICS"
        onClick={[Function]}
        style={
          Object {
            "border-color": "#353535",
            "color": "#353535",
          }
        }
      >
        POLITICS
      </span>
    </div>
    <div
      className="flex flex-wrap justify-center w-100 mb3"
    >
      <button
        className="false bg-blue white h3-l h3-m h3-ns h3-xs f3 w-30 bn center"
        onClick={[Function]}
      >
        SAVE
      </button>
    </div>
    <div
      className="flex flex-wrap justify-center w-100"
    >
      <Link
        className="bg-white blue h3-l h3-m h3-ns h3-xs f3 center"
        replace={false}
        to="/news"
      >
        BACK TO HOME
      </Link>
    </div>
  </div>
</div>
`;
